src/cider/include/cider/processor/BatchProcessor.h

```diff
@@ -23,44 +23,70 @@
 #define CIDER_BATCH_PROCESSOR_H
 
 #include <memory>
-#include "BatchProcessorContext.h"
+
+#include "cider/processor/BatchProcessorContext.h"
 #include "cider/processor/JoinHashTableBuilder.h"
-#include "include/cider/batch/CiderBatch.h"
 #include "substrait/plan.pb.h"
 
-namespace cider::processor {
+struct ArrowArray;
+struct ArrowSchema;
+
+namespace cider::exec::processor {
 
 enum class BatchProcessorState {
   kRunning,
   kWaiting,
   kFinished,
 };
 
-class BatchProcessor {
+class BatchProcessor : public std::enable_shared_from_this<BatchProcessor> {
  public:
-  virtual BatchProcessorContextPtr context() = 0;
+  enum class Type {
+    kStateless,
+    kStateful,
+  };
+
+  virtual const BatchProcessorContextPtr& context() const = 0;
   /// Adds an input batch to the batchProcessor.  This method will only be called if
   /// getState return kRunning.
-  virtual void processNextBatch(std::shared_ptr<CiderBatch> batch) = 0;
+  virtual void processNextBatch(const struct ArrowArray* array,
+                                const struct ArrowSchema* schema = nullptr) = 0;
 
   /// Gets an output batch from the batchProcessor.  return null If no output data.
-  virtual std::shared_ptr<CiderBatch> getResult() = 0;
+  virtual void getResult(struct ArrowArray& array, struct ArrowSchema& schema) = 0;
 
   /// Notifies the batchProcessor that no more batch will be added and the
   /// batchProcessor should finish processing and flush results.
   virtual void finish() = 0;
 
   virtual BatchProcessorState getState() = 0;
 
+  virtual Type getType() const = 0;
```

Comment:
getProcessorType. I think it's necessary to provide such API, user could check processor type for a given plan.